src/pke/lib/scheme/ckksrns/ckksrns-leveledshe.cpp

```diff
@@ -435,41 +472,71 @@ std::vector<DCRTPoly::Integer> LeveledSHECKKSRNS::GetElementForEvalMult(
           ciphertext->GetCryptoParameters());
 
   const std::vector<DCRTPoly> &cv = ciphertext->GetElements();
+  uint32_t numTowers = cv[0].GetNumOfElements();
+  std::vector<DCRTPoly::Integer> moduli(numTowers);
+  for (usint i = 0; i < numTowers; i++) {
+    moduli[i] = cv[0].GetElementAtIndex(i).GetModulus();
+  }
+
   double scFactor = cryptoParams->GetScalingFactorReal(ciphertext->GetLevel());
 
 #if defined(HAVE_INT128)
   typedef int128_t DoubleInteger;
+  int32_t MAX_BITS_IN_WORD = 126;
 #else
   typedef int64_t DoubleInteger;
+  int32_t MAX_BITS_IN_WORD = 62;
 #endif
 
-  DoubleInteger large = static_cast<DoubleInteger>(constant * scFactor + 0.5);
+  // Compute approxFactor, a value to scale down by, in case the value exceeds a 64-bit integer.
+  int32_t logSF = static_cast<int32_t>(ceil(log2(fabs(scFactor))));
+  int32_t logValid = (logSF <= MAX_BITS_IN_WORD) ? logSF : MAX_BITS_IN_WORD;
+  int32_t logApprox = logSF - logValid;
+  double approxFactor = pow(2, logApprox);
+
+  DoubleInteger large = static_cast<DoubleInteger>(constant / approxFactor * scFactor + 0.5);
   DoubleInteger large_abs = (large < 0 ? -large : large);
   DoubleInteger bound = (uint64_t)1 << 63;
 
-  uint32_t numTowers = cv[0].GetNumOfElements();
   std::vector<DCRTPoly::Integer> factors(numTowers);
 
   if (large_abs > bound) {
     for (usint i = 0; i < numTowers; i++) {
-      DCRTPoly::Integer modulus = cv[0].GetElementAtIndex(i).GetModulus();
-      DoubleInteger reduced = large % modulus.ConvertToInt();
+      DoubleInteger reduced = large % moduli[i].ConvertToInt();
 
       factors[i] = (reduced < 0) ?
-        static_cast<uint64_t>(reduced + modulus.ConvertToInt()) :
+        static_cast<uint64_t>(reduced + moduli[i].ConvertToInt()) :
         static_cast<uint64_t>(reduced);
     }
   } else {
     int64_t scConstant = static_cast<int64_t>(large);
     for (usint i = 0; i < numTowers; i++) {
-      DCRTPoly::Integer modulus = cv[0].GetElementAtIndex(i).GetModulus();
-      int64_t reduced = scConstant % static_cast<int64_t>(modulus.ConvertToInt());
+      int64_t reduced = scConstant % static_cast<int64_t>(moduli[i].ConvertToInt());
 
       factors[i] = (reduced < 0) ?
-          reduced + modulus.ConvertToInt() :
+          reduced + moduli[i].ConvertToInt() :
           reduced;
     }
   }
+
+  // Scale back up by approxFactor within the CRT multiplications.
+  int32_t MAX_LOG_STEP = 60;
+  if (logApprox > 0) {
+    int32_t logStep = (logApprox <= MAX_LOG_STEP) ? logApprox : MAX_LOG_STEP;
+    DCRTPoly::Integer intStep = uint64_t(1) << logStep;
+    std::vector<DCRTPoly::Integer> crtApprox(numTowers, intStep);
+    logApprox -= logStep;
+
+    while (logApprox > 0) {
```

Comment:
infinite loop if (logApprox > 0)

Reply:
fixed. There was a typo that had logSF instead of logApprox